### PR TITLE
Issue #146 - Specify GREASE algorithms (aka arbitrary brand and version number)

### PR DESCRIPTION
**Pull request by miketaylr**
PTAL @amtunlimited 

(Right now I'm running a crawler to ensure `_ ( ) ; ? = - : . /` are safe to use, but so far it's looking good).

This also should address https://github.com/WICG/ua-client-hints/issues/161.

***
<a href="https://pr-preview.s3.amazonaws.com/miketaylr/ua-client-hints/pull/178.html" title="Last updated on Jan 7, 2021, 5:35 PM UTC (f9e8cf5)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/ua-client-hints/178/85c1b52...miketaylr:f9e8cf5.html" title="Last updated on Jan 7, 2021, 5:35 PM UTC (f9e8cf5)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -411,49 +411,89 @@ Note: The high-entropy portions of the user agent information are retrieved thro
...
 3. For each pair:
 
     1. Let |dict| be a new {{NavigatorUABrandVersion}} dictionary,
-        with [=user agent/brand=] as {{NavigatorUABrandVersion/brand}} and [=user agent/significant version=] as {{NavigatorUABrandVersion/version}}.
+        with [=user agent/brand=] as {{NavigatorUABrandVersion/brand}} and [=user agent/significant version=] as
+        {{NavigatorUABrandVersion/version}}.
 
     2. Append |dict| to |list|.
 
 4.  The [=user agent=] SHOULD execute the following steps:
```
Comment (amtunlimited) [inline index.bs:439]: Somewhat unfamiliar with spec language; does this wording imply that if you don't do one, you shouldn't do the other?

File: index.bs [hunk 2]
```diff
@@ -411,49 +411,89 @@ Note: The high-entropy portions of the user agent information are retrieved thro
...
-<h3 id="getters">Getters</h3>
+An <dfn for="user agent" export>equivalence class</dfn> represents a group of browsers believed to be compatibile with
+each other. A shared rendering engine may form an [=equivalence class=], for example.
+
+<h4 id="create-arbitrary-brands-section"
+    algorithm="to create arbitrary brand and version values">Create arbitrary brand and version values</h4>
+
+To <dfn>create an arbitrary brand</dfn>, the [=user agent=] MUST run these steps:
+
+    1. Let |arbitraryBrand| be a [=/string=] composed of [=ASCII lower alpha=], [=ASCII upper alpha=], or 0x20 (SP), no
```
Comment (amtunlimited) [inline index.bs:463]: Is the char-set or byte limit imposed on normal brands as well? Should it be?

File: index.bs [hunk 3]
```diff
@@ -411,49 +411,89 @@ Note: The high-entropy portions of the user agent information are retrieved thro
...
+         longer than twenty [=ASCII bytes=].
+    1. Let |arbitraryBrandList| be the result of <a lt="split on ASCII whitespace">splitting |arbitraryBrand| on ASCII
```
Comment (amtunlimited) [inline index.bs:465]: Should there be a minimum number of spaces included to ensure there are a minimum number of GREASEy chars?

File: index.bs (hunk 2 above)
Comment (amtunlimited) [inline index.bs:463]: Should we include numbers in this range?

Comment (amtunlimited) [inline index.bs:463]: (hat trick!) nit: s/`[=ASCII lower alpha=], [=ASCII upper alpha=]`/`[=ASCII alpha=]`/ (or [=ASCII alphanumeric=] if we decide to add numbers into the mix)

File: index.bs [hunk 4]
```diff
@@ -411,49 +411,89 @@ Note: The high-entropy portions of the user agent information are retrieved thro
...
+    1. Let |greaseyChars| be the [=list=] of [=ASCII bytes=] « 0x20 (SP), 0x28 (left parenthesis), 0x29
+         (right parenthesis), 0x2D (-), 0x2E (.), 0x2F (/), 0x3A (:), 0x3B (;), 0x3D (=), 0x3F (?), 0x5F (_) ».
+    1. For each item of |arbitraryBrandList|, [=stack/push=] a randomly selected [=list/item=] from |greaseyChars| onto
+         |greaseyStack|.
+    1. Let |greaseyBrandList| be a [=list=] and |index| be 0.
+    1. While |greaseyStack| [=stack/is not empty=]:
+        1. Let |item| be the result of <a lt="pop">popping</a> from |greaseyStack|.
+        1. Append |item| to |greaseyBrandList|.
+        1. Append |arbitraryBrandList|[|index|] to |greaseyBrandList|.
+        1. Increment |index| by 1.
```
Comment (amtunlimited) [inline index.bs:477]: Might this end up in an array-out-of-bounds?

File: index.bs (hunk 1 above)
Comment (miketaylr) [reply index.bs:439]: Yes, if you were to only pick one, you would be violating the spec.

File: index.bs (hunk 2 above)
Comment (miketaylr) [reply index.bs:463]: It's not, but we can open an issue to consider it: https://github.com/WICG/ua-client-hints/issues/179

File: index.bs (hunk 3 above)
Comment (miketaylr) [reply index.bs:465]: That's an interesting point. If we require at least 1 space, you end up with at least 2 greasy chars. WDYT @amtunlimited?

File: index.bs (hunk 2 above)
Comment (miketaylr) [reply index.bs:463]: I think we shouldn't. I don't think there's a good reason to encourage parsing of numbers in a brand name, instead consumers should look in the version field.

(And I double checked that there's nothing surprising here:
https://en.wikipedia.org/wiki/List_of_web_browsers)

Comment (miketaylr) [reply index.bs:463]: Oh nice, thanks!

File: index.bs (hunk 3 above)
Comment (miketaylr) [reply index.bs:465]: (I think I'm going to make 1 space a requirement, but open to other ideas)

File: index.bs (hunk 4 above)
Comment (miketaylr) [reply index.bs:477]: I don't see how, do you have a scenario where it might?

Comment (amtunlimited) [reply index.bs:477]: I misread step 5; silly me

**Review by amtunlimited — APPROVED** [review]
LGTM

**Comment by miketaylr** [comment]
Thanks for the review!

**Pull request merged**